File: Backend/app/domain/entities/detection.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import List, Optional
from app.domain.entities.location import Location


@dataclass
class DetectionItem:
    id: Optional[str]
    class_name: str
    waste_group: str
    confidence: float
    bbox_x: float
    bbox_y: float
    bbox_w: float
    bbox_h: float


@dataclass
class Detection:
    id: Optional[str]
    image_url: str
    annotated_image_url: Optional[str] = None
    processing_time_ms: Optional[int] = None
    model_version: str = "v1.0"
    detection_status: str = "pending"
    failure_reason: Optional[str] = None
    location: Optional[Location] = None
    location_source: Optional[str] = None
    location_confidence: Optional[float] = None
    items: List[DetectionItem] = field(default_factory=list)
    created_at: Optional[datetime] = field(default_factory=lambda: datetime.now(timezone.utc))

File: Frontend/src/types/detection.ts
// ─── Detection Item ──────────────────────────────────────────────

export interface DetectionItem {
  id: string;
  waste_type: string;
  confidence: number;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
}

// ─── Detection Response ──────────────────────────────────────────

export interface Detection {
  id: string;
  image_url: string;
  latitude: number;
  longitude: number;
  model_version: string;
  detection_status: 'pending' | 'completed' | 'failed';
  failure_reason: string | null;
  items: DetectionItem[];
  created_at: string;
}

// ─── API Envelope ────────────────────────────────────────────────

export interface DetectionEnvelope {
  data: Detection;
  meta: Record<string, unknown> | null;
  error: string | null;
}

// ─── Risk Assessment ─────────────────────────────────────────────

export interface RiskBreakdown {
  object_count: number;
  density: number;
  hazard: number;
  waterbody: number;
}

export interface RiskAssessment {
  id: string;
  detection_id: string;
  score: number;
  level: 'low' | 'medium' | 'high' | 'critical';
  strategy_breakdown: RiskBreakdown;
  computed_at: string;
}

export interface RiskEnvelope {
  data: RiskAssessment;
  meta: Record<string, unknown> | null;
  error: string | null;
}
